homeassistant/components/recorder/queries.py

```diff
@@ -0,0 +1,462 @@
+"""Queries for the recorder."""
+from __future__ import annotations
+
+from sqlalchemy import func, lambda_stmt, select, union_all
+from sqlalchemy.sql.lambdas import StatementLambdaElement
+from sqlalchemy.sql.selectable import Select
+
+from .models import EventData, Events, StateAttributes, States
+
+
+def find_shared_attributes_id(
+    data_hash: int, shared_attrs: str
+) -> StatementLambdaElement:
+    """Find an attributes_id by hash and shared_attrs."""
+    return lambda_stmt(
+        lambda: select(StateAttributes.attributes_id)
+        .filter(StateAttributes.hash == data_hash)
+        .filter(StateAttributes.shared_attrs == shared_attrs)
+    )
+
+
+def find_shared_data_id(attr_hash: int, shared_data: str) -> StatementLambdaElement:
+    """Find a data_id by hash and shared_data."""
+    return lambda_stmt(
+        lambda: select(EventData.data_id)
+        .filter(EventData.hash == attr_hash)
+        .filter(EventData.shared_data == shared_data)
+    )
+
+
+def _state_attrs_exist(attr: int | None) -> Select:
+    """Check if a state attributes id exists in the states table."""
+    return select(func.min(States.attributes_id)).where(States.attributes_id == attr)
+
+
+def attributes_ids_exist_in_states(
```

Comment:
While this performs and works well, I wish we could make these shorter, but I haven't found a way after many hours of working with `lambda_stmt`.